src/Sarif.WorkItems/SarifWorkItemsExtensions.cs

```diff
@@ -0,0 +1,89 @@
+{
+    public static class SarifWorkItemsExtensions
+    {
+        public static bool AppropriateForFiling(this Result result)
+        {
+            // Fail: an explicit failure occurred.
+            //
+            // Open: a sound analysis is 'open' indicating that the analysis requires
+            //       more configuration or other data in order to produce a determination.
+            // 
+            // Review: an open item which can't be automated, i.e., which requires a
+            //         manual review to resolve.
+            return result.Kind == ResultKind.Fail ||
+                   result.Kind == ResultKind.Open ||
+                   result.Kind == ResultKind.Review;
+
+            // Designations which are not appropriate for filing:
+            //
+            // Pass: the scan target explicitly passed analysis.
+            // 
+            // Not applicable: analysis was skipped because it was not
+            //                 relevant to the scan target.
+            //
+            // Informational: a strictly informational message was emitted.
+        }
+
+        public static string CreateWorkItemTitle(this Run run)
+        {
+            Result firstResult = null;
+            var ruleIds = new HashSet<string>();
+
+            foreach (Result result in run?.Results)
+            {
+                if (!result.AppropriateForFiling())
+                {                     
+                    continue; 
+                }
+
+                string ruleId = result.ResolvedRuleId(run);
+                firstResult = firstResult ?? result;
+                ruleIds.Add(ruleId);
+            }
+
+            // No useful work to schedule in a work item, apparently.
+            if (firstResult == null) { return null; }
+
+            string fullRuleId = ConstructFullRuleIdentifier(firstResult.GetRule(run));
+
+            // We encapsulate logical names in apostrophes to help indicate they are a symbol
+            string locationName = "'" + firstResult.Locations?[0].LogicalLocation?.FullyQualifiedName + "'";
+
+            if (locationName.Equals("''"))
+            {
+                // We don't bother to wrap a file path or URL in apostrophes as these are self-evident
+                //
+                // Lines of code like this that inspire strong feelings in SARIF consumers as far as its design.
+                locationName = firstResult.Locations?[0].PhysicalLocation?.ArtifactLocation?.Resolve(run)?.Uri.OriginalString;
+            }
+
+            // Returns strings like:
+            //
+            // [Tool:Warning] RULE3067 (in 'Namespace.Type.MyMethod()')
+            // [Tool:Error] RULE2001: Null derefernece (in c:\src\build.cpp)
+
+            return "[" + run.Tool.Driver.Name + ":" +
+                    firstResult.Level.ToString() + "]: " +
+                    fullRuleId +
+                    (locationName == null ? "" : " (in " + locationName + ")");
+        }
```

Comment:
I was looking at the TSA bug title generation code yesterday. Thought I would share in case we want to incorporate any of it.  I think your code is taking the title in a slightly different direction. There are folks though who already have post processing hooks looking for this pattern. 

GlobalConfig.Environment is the stamp, it is empty in V2.  

string valueTitle = string.Format(
                "{0}[ {1} ][ {2} ] - {3} ({4})",
                Formatter.FormatIfNotEmpty("( {0} )", GlobalConfig.Environment),
                tsaWorkItem.CodeBase.Name,
                tsaWorkItem.ScanTool.DisplayName,
                tsaWorkItem.IssueGroup,
                Formatter.ToFriendlyQuantity(issueCount, "issue", "issues"));

Reply:
I'm definitely aware of this code. :) Trying to thoughtfully rebuild titles from first principles. I think you will find that the TSA bug bodies argue for something similar. I'm glad you raised this point, though, because it makes me realize we could move some of the old TSA title content into a simplified message body. Wil follow up with you offline. 

---
In reply to: [392506645](https://github.com/microsoft/sarif-sdk/pull/1813#discussion_r392506645) [](ancestors = 392506645)